File: exercism/javascript/beer-song/beer-song.js
const { mapValues, last } = require('lodash/fp')

const bridge = [
  'Go to the store and buy some more',
  'Take it down and pass it around',
  'Take one down and pass it around'
]

const format = ({ n, nUpper, m, pluralN, pluralM, bridge}) =>
  `${nUpper} bottle${pluralN} of beer on the wall, ${n} bottle${pluralN} of beer.
${bridge}, ${m} bottle${pluralM} of beer on the wall.
`

class BeerSong {
  verse (n) {
    const params = mapValues(xs => xs[n] || last(xs), {
      n: ['no more', n],
      nUpper: ['No more', n],
      m: [99, 'no more', n - 1],
      pluralN: ['s', '', 's'],
      pluralM: ['s', 's', '', 's'],
      bridge: bridge
    })
    return format(params)
  }
  sing (a, b) {
    return range(a, b).map(verse).join('\n\n')
  }
}

module.exports = BeerSong

File: exercism/cpp/beer-song/beer_song.cpp
#include "beer_song.h"
#include <iostream>
#include <sstream>

std::string beer::verse(int n) {
  switch (n) {
  case 0:
    return "No more bottles of beer on the wall, no more bottles of beer.\nGo "
           "to the store and buy some more, 99 bottles of beer on the wall.\n";
  case 1:
    return "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and "
           "pass it around, no more bottles of beer on the wall.\n";
  case 2:
    return "2 bottles of beer on the wall, 2 bottles of beer.\nTake one down "
           "and pass it around, 1 bottle of beer on the wall.\n";
  default:
    std::stringstream ss;
    ss << n << " bottles of beer on the wall, " << n
       << " bottles of beer.\nTake one down and pass it around, " << n - 1
       << " bottles of beer on the wall.\n";
    return ss.str();
  }
}

std::string beer::sing(int start, int end) {
  std::stringstream ss;
  for (int i = start; i >= end; i--) {
    if (i != start)
      ss << '\n';
    ss << verse(i);
  }
  return ss.str();
}
